### PR TITLE
[Fix] Wrong permissions on edit CDB

### PR DESCRIPTION
**Pull request by gabiwassan**
Hi team,

This PR resolves:
- Wrong permissions on edit CDB List with Cluster and Manager context configuration server.
- Added action to get context server and status.
- Context server is dynamically added to the CDBList edit button, add new, and delete.
Note: Policy `cluster:status` is necessary to get configuration context `manager` or `cluster`

Issue #2657

### CODE REVIEW

File: public/redux/reducers/appStateReducers.js [hunk 1]
```diff
@@ -120,6 +120,13 @@ const appStateReducers = (state = initialState, action) => {
     };
   }
 
+  if (action.type === 'UPDATE_CLUSTER_STATUS') {
+    return {
+      ...state,
+      clusterStatus: action.clusterStatus,
```
Comment (frankeros) [inline public/redux/reducers/appStateReducers.js:126]: you are missing add to initial state the prop `clusterStatus`  for this reducer

**Review by frankeros — CHANGES_REQUESTED** [review]

File: public/redux/reducers/appStateReducers.js (hunk 1 above)
Comment (gabiwassan) [reply public/redux/reducers/appStateReducers.js:130]: Thx!!! 😄

File: public/controllers/management/components/management/ruleset/ruleset-table.js [hunk 2]
```diff
@@ -122,14 +125,41 @@ class WzRulesetTable extends Component {
     this._isMounted = false;
   }
 
+  isClusterOrManager = async () => {
+    await clusterReq()
+      .then((clusterStatus) => {
```
[the remark this answers is not in the document]
Comment (gabiwassan) [reply public/controllers/management/components/management/ruleset/ruleset-table.js:130]: Are you sure? I think you cannot use try-catch statements to handle exceptions thrown asynchronously, as the function has "returned" before any exceptions are thrown.

File: public/controllers/management/components/management/ruleset/actions-buttons.js [hunk 3]
```diff
@@ -199,59 +199,79 @@ class WzRulesetActionButtons extends Component {
...
-    const addNewCdbListButton = (
-      <WzButtonPermissions
-        buttonType='empty'
-        permissions={[
-          {
-            action: `${((this.props || {}).clusterStatus || {}).contextConfigServer}:upload_file`,
-            resource: 'file:path:/etc/lists/files',
-          },
+    const getPermissionsNewFile = () => {
+      let permissions = [
```
Comment (frankeros) [inline public/controllers/management/components/management/ruleset/actions-buttons.js:228]: `permissions` can be a `const`

File: public/controllers/management/components/management/ruleset/ruleset-table.js [hunk 4]
```diff
@@ -236,29 +236,45 @@ class WzRulesetTable extends Component {
...
-          'data-test-subj': `row-${id || name}`,
-          className: 'customRowClass',
-          onClick: !WzUserPermissions.checkMissingUserPermissions(
-            [
-              {
-                action: `${((this.props || {}).clusterStatus || {}).contextConfigServer}:read_file`,
-                resource: `file:path:${item.relative_dirname}/${item.filename}`,
-              },
+        const getRequiredPermissions = (item) => {
+          let permissions = [
```
Comment (frankeros) [inline public/controllers/management/components/management/ruleset/ruleset-table.js:240]: same here

File: public/controllers/management/components/management/ruleset/utils/columns.js [hunk 5]
```diff
@@ -249,21 +249,39 @@ export default class RulesetColumns {
       };
 
       const getEditButtonPermissions = (item) => {
-        return [
+        let permissions = [
```
Comment (frankeros) [inline public/controllers/management/components/management/ruleset/utils/columns.js:252]: same

**Review by frankeros — CHANGES_REQUESTED** [review]

File: public/controllers/management/components/management/ruleset/actions-buttons.js (hunk 3 above)
Comment (gabiwassan) [reply public/controllers/management/components/management/ruleset/actions-buttons.js:228]: Sure! thanks

File: public/react-services/wz-user-permissions.ts [hunk 6]
```diff
@@ -950,8 +951,32 @@ export class WzUserPermissions{
...
+              ? ![actionResource, actionResourceAll].includes(resource) &&
+                (resource.match(`/${actionResource}/`) || resource.match(`/${actionResourceAll}/`))
+              : ![actionResource, actionResourceAll].includes(resource) &&
+                (resource.match(actionResource.replace('*', '\\*')) ||
+                  resource.match(actionResourceAll.replace('*', '\\*')))
+          )
+        : undefined;
+
+      userPartialResources =
+        (userPartialResources || []).length > 0 ? userPartialResources : undefined;
```
Comment (frankeros) [inline public/react-services/wz-user-permissions.ts:978]: you can save a couple code lines with 
```js
      const userPartialResources: string[] | undefined = Object.keys(userPermissions[actionName] || {}).length
        ? Object.keys(userPermissions[actionName]).filter((resource) =>
            resource.match('&')
              ? ![actionResource, actionResourceAll].includes(resource) &&
                (resource.match(`/${actionResource}/`) || resource.match(`/${actionResourceAll}/`))
              : ![actionResource, actionResourceAll].includes(resource) &&
                (resource.match(actionResource.replace('*', '\\*')) ||
                  resource.match(actionResourceAll.replace('*', '\\*')))
          )
        : undefined;
```

**Review by frankeros — CHANGES_REQUESTED** [review]

File: public/react-services/wz-user-permissions.ts (hunk 6 above)
Comment (gabiwassan) [reply public/react-services/wz-user-permissions.ts:978]: Excellet! let me try and test 😄

**Review by jsanchez91 — CHANGES_REQUESTED** [review]
I have conducted tests with these users and roles using X-Pack Security:
- **elastic**: administrator
- **test_user**: readonly, cluster_readonly

This is the result I have obtained in both tests

image

**Comment by gabiwassan** [comment]
> I have conducted tests with these users and roles using X-Pack Security:
> 
> * **elastic**: administrator
> * **test_user**: readonly, cluster_readonly
> 
> This is the result I have obtained in both tests
> 
> image


Thanks, I'm checking it right now.

**Comment by frankeros** [comment]
Hi @gabiwassan 

#### Testing this branch I found this bug when you set these policies to the role
image

You're able to see all files but trying to access one of them throw this error
image

The same error occurs 
image

----

#### Another bug that I found when you have a specific path in `file:path`
like this 
image

You see it as disabled
image

----

#### Also with the same policies, when you try to add new lists 
image
It requires a specific path (I'm not sure if it is a bug)
Screenshot from 2020-12-10 10-02-26

File: public/controllers/management/components/management/ruleset/actions-buttons.js [hunk 7]
```diff
@@ -214,8 +301,25 @@ class WzRulesetActionButtons extends Component {
...
-        permissions={[{action: 'manager:upload_file', resource: `file:path:/etc/${section}`}]}
+        buttonType="empty"
+        permissions={[
+          {
+            action: `${((this.props || {}).clusterStatus || {}).contextConfigServer}:upload_file`,
+            resource: `file:path:/etc/${section}`,
+          },
+          {
+            action: `${((this.props || {}).clusterStatus || {}).contextConfigServer}:read`,
+            resource: `file:path:/etc/${this.props.msg}`,
```
Comment (Desvelao) [inline public/controllers/management/components/management/ruleset/actions-buttons.js:312]: Should it be the `section` variable instead `this.props.msg`? I checked the code and didn't found where it is coming from

File: public/controllers/management/components/management/ruleset/actions-buttons.js [hunk 8]
```diff
@@ -214,8 +301,25 @@ class WzRulesetActionButtons extends Component {
...
+          },
+          {
+            action: `${((this.props || {}).clusterStatus || {}).contextConfigServer}:read_file`,
+            resource: `file:path:/etc/${this.props.msg}`,
```
Comment (Desvelao) [inline public/controllers/management/components/management/ruleset/actions-buttons.js:316]: Should it be the `section` variable instead `this.props.msg`? I checked the code and didn't found where it is coming from

**Comment by Desvelao** [comment]
I tested the PR and found the next problem:
- Logged with deafult `admin` user in OpenDistro, I foud this problem in the `Management/Rules` section:
image

**Comment by Desvelao** [comment]
I don't think we need check the permissions here, allowing edit/remove items of the list.
image
image

I think only the `Save button` is enough to do the validation to upload the file.

I found a problem using the default `admin` user that has the default `administrator` role. The `Save button` is disabled because the required `resource` is not included in the `action` that has this role.

image
image

Maybe modify the `resource` to `node:id:*` (Although the file should really have to be uploaded to the master node of the cluster, i mean the resource should be `node:id:<master_node_id>` instead `node:id:*` to be more exactly)

**Comment by frankeros** [comment]
Testing the new commits I found these two bugs

- When the user has permissions to a specific file and permissions to `upload_file`
image
and try to save a new list file gets this error
image
but the new item is created anyway

- When the user (any with access to this action) tries to export a `list` detail, gets this error
image
```
code: 3034
message: "3034 - An error occurred fetching data from the Wazuh API : Extra query parameter(s) path not in spec"
statusCode: 500
```

**Review by frankeros — APPROVED** [review]
great job @gabiwassan!! :clap: :clap:

**Comment by gabiwassan** [comment]
> Testing the new commits I found these two bugs
> 
> * When the user has permissions to a specific file and permissions to `upload_file`
>   image
>   and try to save a new list file gets this error
>   image
>   but the new item is created anyway
> * When the user (any with access to this action) tries to export a `list` detail, gets this error
>   image
> 
> ```
> code: 3034
> message: "3034 - An error occurred fetching data from the Wazuh API : Extra query parameter(s) path not in spec"
> statusCode: 500
> ```

The following issue was created https://github.com/wazuh/wazuh-kibana-app/issues/2692 for error 500

**Review by Desvelao — APPROVED** [review]
LGTM

**Pull request merged**